### PR TITLE
Support LaunchTemplates by Name

### PR DESCRIPTION
**Pull request by dnatic09**
LaunchTemplates cannot be assigned to NodeGroups by Name. This commit allows for assignment via ID or Name

### Description
Support of assigning Launch Templates to NodeGroups via Name

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

### My Comments
I believe this enhancement will help with teams that would like to control more of the infrastructure stack in CloudFormation. It is easy to define a LaunchTemplate in CloudFormation and then feed that name into an eksctl configuration. Using ID is restrictive because it is dynamically generated upon creation of the LaunchTemplate resource.

### CODE REVIEW

**Comment by dnatic09** [comment]
Good afternoon eksctl Team,

I would like to write more unit tests. I am not quite able to reverse engineer the pattern. I would like to mock `DescribeLaunchTemplate` and `DescribeLaunchTemplateVersions` to better reflect possibilities of responses. Any pointers on which files and pieces to modify would be great.

**Comment by dnatic09** [comment]
Manual Test Results succeeded:

# Input
```yaml
#ClusterConfig object using an existing VPC:
---
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: poc1-k8s-cluster
  region: eu-west-1
  version: '1.24'

vpc:
  id: vpc-98e140fd  # (optional, must match VPC ID used for each subnet below)
  clusterEndpoints:
    privateAccess: true
    publicAccess: false
  sharedNodeSecurityGroup: sg-bc4239d9
  subnets:
    # must provide 'private' subnets by availability zone as shown
    private:
      dev1-private-subnet-a:
        id: subnet-73cd5016
      dev1-private-subnet-b:
        id: subnet-8f72c8f8
      dev1-private-subnet-c:
        id: subnet-e8fa25b1


addons:
  - name: vpc-cni
  - name: kube-proxy
  - name: coredns

managedNodeGroups:
  - name: nodegroup-1
    minSize: 1
    maxSize: 2
    desiredCapacity: 2
    instanceType: m5.large
    volumeSize: 50
    privateNetworking: true
    ssh:
      allow: true
      publicKeyName: dev
    labels: {role: worker}
    tags:
      nodegroup-role: worker
      k8s.io/cluster-autoscaler/poc1-k8s-cluster: owned
      k8s.io/cluster-autoscaler/enabled: 'true'
    securityGroups:
      attachIDs:
        - sg-bc4239d9
    subnets:
      - dev1-private-subnet-a
      - dev1-private-subnet-b
      - dev1-private-subnet-c

  - name: ng-test-dan-1
    minSize: 1
    maxSize: 1
    desiredCapacity: 1
    privateNetworking: true
    launchTemplate:
#      id: lt-025eaaa8b41bb970b
      name: eksctl-dev1-k8s-cluster-nodegroup-nodegroup-1
      version: "2"
```

# Console Output
```
2022-12-28 20:21:35 [_]  eksctl version 0.125.0-dev+94d489e59.2022-12-28T18:58:44Z
2022-12-28 20:21:35 [_]  using region eu-west-1
2022-12-28 20:21:35 [!]  warning, having public access disallowed will subsequently interfere with some features of eksctl. This will require running subsequent eksctl (and Kubernetes) commands/API calls from within the VPC.  Running these in the VPC requires making updates to some AWS resources.  See: https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html for more details
2022-12-28 20:21:35 [!]  security group rules may be added by eksctl; see vpc.manageSharedNodeSecurityGroupRules to disable this behavior
2022-12-28 20:21:37 [_]  using existing VPC (vpc-98e140fd) and subnets (private:map[dev1-private-subnet-a:{subnet-73cd5016 eu-west-1a 10.188.64.0/21 0 } dev1-private-subnet-b:{subnet-8f72c8f8 eu-west-1b 10.188.72.0/21 0 } dev1-private-subnet-c:{subnet-e8fa25b1 eu-west-1c 10.188.80.0/21 0 }] public:map[])
2022-12-28 20:21:37 [!]  custom VPC/subnets will be used; if resulting cluster doesn't function as expected, make sure to review the configuration of VPC/subnets
2022-12-28 20:21:37 [_]  nodegroup "nodegroup-1" will use "" [AmazonLinux2/1.24]
2022-12-28 20:21:37 [_]  using EC2 key pair "dev"
2022-12-28 20:21:37 [_]  nodegroup "ng-test-dan-1" will use "" [AmazonLinux2/1.24]
2022-12-28 20:21:37 [_]  using Kubernetes version 1.24
2022-12-28 20:21:37 [_]  creating EKS cluster "poc1-k8s-cluster" in "eu-west-1" region with managed nodes
2022-12-28 20:21:37 [_]  2 nodegroups (ng-test-dan-1, nodegroup-1) were included (based on the include/exclude rules)
2022-12-28 20:21:37 [_]  will create a CloudFormation stack for cluster itself and 0 nodegroup stack(s)
2022-12-28 20:21:37 [_]  will create a CloudFormation stack for cluster itself and 2 managed nodegroup stack(s)
2022-12-28 20:21:37 [_]  if you encounter any issues, check CloudFormation console or try 'eksctl utils describe-stacks --region=eu-west-1 --cluster=poc1-k8s-cluster'
2022-12-28 20:21:37 [_]  Kubernetes API endpoint access will use provided values {publicAccess=false, privateAccess=true} for cluster "poc1-k8s-cluster" in "eu-west-1"
2022-12-28 20:21:37 [_]  CloudWatch logging will not be enabled for cluster "poc1-k8s-cluster" in "eu-west-1"
2022-12-28 20:21:37 [_]  you can enable it with 'eksctl utils update-cluster-logging --enable-types={SPECIFY-YOUR-LOG-TYPES-HERE (e.g. all)} --region=eu-west-1 --cluster=poc1-k8s-cluster'
2022-12-28 20:21:37 [_]
2 sequential tasks: { create cluster control plane "poc1-k8s-cluster",
    2 sequential sub-tasks: {
        2 sequential sub-tasks: {
            wait for control plane to become ready,
            1 task: { create addons },
        },
        2 parallel sub-tasks: {
            create managed nodegroup "nodegroup-1",
            create managed nodegroup "ng-test-dan-1",
        },
    }
}
2022-12-28 20:21:37 [_]  building cluster stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:21:38 [_]  deploying stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:22:08 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:22:39 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:23:40 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:24:41 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:25:42 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:26:42 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:27:43 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:28:44 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:29:45 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:30:46 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:31:47 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:32:48 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:33:48 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:34:49 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-cluster"
2022-12-28 20:36:56 [!]  OIDC is disabled but policies are required/specified for this addon. Users are responsible for attaching the policies to all nodegroup roles
2022-12-28 20:36:56 [_]  creating addon
2022-12-28 20:37:07 [_]  addon "vpc-cni" active
2022-12-28 20:37:10 [_]  building managed nodegroup stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:37:10 [_]  building managed nodegroup stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:37:11 [_]  deploying stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:37:12 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:37:12 [_]  deploying stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:37:12 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:37:43 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:37:43 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:38:30 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:38:39 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:39:22 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:39:59 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-ng-test-dan-1"
2022-12-28 20:41:04 [_]  waiting for CloudFormation stack "eksctl-poc1-k8s-cluster-nodegroup-nodegroup-1"
2022-12-28 20:41:04 [_]  waiting for the control plane to become ready
2022-12-28 20:41:07 [!]  unable to write kubeconfig , please retry with 'eksctl utils write-kubeconfig -n poc1-k8s-cluster': unable to modify kubeconfig /root/.kube/config: open /root/.kube/config.lock: permission denied
2022-12-28 20:41:07 [_]  no tasks
2022-12-28 20:41:07 [_]  all EKS cluster resources for "poc1-k8s-cluster" have been created
2022-12-28 20:41:08 [_]  nodegroup "nodegroup-1" has 2 node(s)
2022-12-28 20:41:08 [_]  node "ip-10-188-71-252.eu-west-1.compute.internal" is ready
2022-12-28 20:41:08 [_]  node "ip-10-188-76-86.eu-west-1.compute.internal" is ready
2022-12-28 20:41:08 [_]  waiting for at least 1 node(s) to become ready in "nodegroup-1"
2022-12-28 20:41:08 [_]  nodegroup "nodegroup-1" has 2 node(s)
2022-12-28 20:41:08 [_]  node "ip-10-188-71-252.eu-west-1.compute.internal" is ready
2022-12-28 20:41:08 [_]  node "ip-10-188-76-86.eu-west-1.compute.internal" is ready
2022-12-28 20:41:09 [_]  nodegroup "ng-test-dan-1" has 1 node(s)
2022-12-28 20:41:09 [_]  node "ip-10-188-64-207.eu-west-1.compute.internal" is ready
2022-12-28 20:41:09 [_]  waiting for at least 1 node(s) to become ready in "ng-test-dan-1"
2022-12-28 20:41:09 [_]  nodegroup "ng-test-dan-1" has 1 node(s)
2022-12-28 20:41:09 [_]  node "ip-10-188-64-207.eu-west-1.compute.internal" is ready
2022-12-28 20:41:11 [_]  creating addon
2022-12-28 20:41:22 [_]  addon "kube-proxy" active
2022-12-28 20:41:22 [_]  creating addon
2022-12-28 20:41:33 [_]  addon "coredns" active
2022-12-28 20:41:40 [_]  EKS cluster "poc1-k8s-cluster" in "eu-west-1" region is ready
Create iamidentitymapping for admin-cluster role if it wasn't created yet ...
2022-12-28 20:41:48 [_]  checking arn arn:aws:iam::018471812555:role/poc1-k8s-cluster-EKSAdminUserRole against entries in the auth ConfigMap
2022-12-28 20:41:48 [_]  adding identity "arn:aws:iam::018471812555:role/poc1-k8s-cluster-EKSAdminUserRole" to auth ConfigMap
Create iamidentitymapping for view role if it wasn't created yet ...

No differences detected in /root/ws/av-stack-init/cloudformation/eks-cluster/files/viewClusterRoleBinding.yaml.
clusterrolebinding.rbac.authorization.k8s.io/read-all-resources unchanged
2022-12-28 20:42:10 [_]  checking arn arn:aws:iam::018471812555:role/poc1-k8s-cluster-EKSViewUserRole against entries in the auth ConfigMap
2022-12-28 20:42:10 [_]  adding identity "arn:aws:iam::018471812555:role/poc1-k8s-cluster-EKSViewUserRole" to auth ConfigMap
```

Screenshot:
image

**Comment by dnatic09** [comment]
Good morning, WeaveWorks Team,

I am happy to finished reviewing my results and tests. This PR seems packaged and ready to merge as-is. I am certainly open to all changes and modifications to have this feature merged in. As stated in the original description, being able to call Launch Templates by name provides an added flexibility that I believe many would utilize in lieu of referencing templates by ID. I look forward to a discussion and review.

**Comment by Himangini** [comment]
thanks for the contribution, we'll review it soon 👍🏻

**Comment by dnatic09** [comment]
@Himangini , thank you for reviewing my PR.

I just did a rebase+force push. I took ALL changes from `main` for `go.mod` and `go.sum`. My branch no longer changes these dependencies. The `main` branch already had more recent dependencies than mine.

**Comment by dnatic09** [comment]
Good morning, @Himangini ,

I noticed my branch is out-of-date against `main`, but will still merge cleanly. I see that all checks have passed with the exception of the PR labels. Should I rebase against `main` (and force-push) or leave as-is?

**Comment by dnatic09** [comment]
Good morning. Just seeing if we can get the unit tests kicked off for this PR.

**Comment by dnatic09** [comment]
Good morning, Team.
I rebased again to catch up to `main`. Any chance we could start the unit tests?

**Comment by Himangini** [comment]
> @Himangini , any chance we can quick off the unit tests?

@dnatic09 oops my bad, I got my hands full with a lot of things so completely lost track of your PR. 
I haven't reviewed it yet but you don't need to rebase/run checks often. We'll run these anyway when we're about to merge. We'll review the PR soon. Thanks for your patience.

**Comment by dnatic09** [comment]
Thanks. Well, the good news is that all tests pass :)

**Comment by dnatic09** [comment]
Just circling back to see if this can be merged, @Himangini .

File: pkg/actions/nodegroup/upgrade.go [hunk 1]
```diff
@@ -373,7 +377,7 @@ func (m *Manager) usesCustomAMIEKSNodeGroup(ctx context.Context, ng *ekstypes.No
 	}
 
 	return m.usesCustomAMI(ctx, &api.LaunchTemplate{
-		ID:      *ng.LaunchTemplate.Id,
+		ID:      ng.LaunchTemplate.Id,
 		Version: ng.LaunchTemplate.Version,
 	})
 }
```
Comment (TiberiuGC) [inline pkg/actions/nodegroup/upgrade.go:383]: I believe the changes in this file are not necessary. Here we first retrieve the launch template for the nodegroup's cloudformation stack, so the resource we get back should always have an ID. Checking for name is redundant in this workflow.

**Review by TiberiuGC — CHANGES_REQUESTED** [review]

File: pkg/actions/nodegroup/upgrade.go (hunk 1 above)
Comment (TiberiuGC) [inline pkg/actions/nodegroup/upgrade.go:383]: I believe the changes in this file are not necessary. Here we first retrieve the launch template from nodegroup's cloudformation stack, hence the resource should always have and ID. Checking for name is redundant in this workflow.

**Comment by Himangini** [comment]
> Hi @dnatic09 , could you please remove all formatting changes that are not related to the outcome of the PR? 🙏 I am seeing a lot of files changed that are only different in formatting.

@dnatic09 Are you able to add the changes requested by @TiberiuGC?

**Pull request closed without merging**